beacon-chain/p2p/peers/scorer.go

```diff
@@ -0,0 +1,116 @@
+package peers
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	// How many bad responses to tolerate before peer is deemed bad.
+	defaultBadResponsesThreshold = 6
+	// Since score represents penalty, it has negative weight.
+	defaultBadResponsesWeight = -1.0
+	// How often to decay previous statistics. Every interval bad responses counter will be decremented by 1.
+	defaultBadResponsesDecayInterval = time.Hour
+)
+
+// PeerScorer keeps track of peer counters that are used to calculate peer score.
+type PeerScorer struct {
+	lock      sync.RWMutex
+	ctx       context.Context
+	params    *PeerScorerParams
+	peerStats map[peer.ID]*peerScorerStats
```

Comment:
probably my main issue with this is that peerScorer is a separate entity from our peer status. We should make the `peerScorerStats` a part of the main peer status state.

Reply:
(copying my response from Discord, so that other reviewers have a bit more context of what I am trying to accomplish):

>Initially, I've implemented scorer w/i status, however, it seemed like I mix two different functionalities violating coherence: that's status is about connectivity to mesh, and scorer is about performance. I am aware that keeping two maps of peers might be inefficient, but we have two distinct components that require it.
>
>Also, peer scorer will become more complex (even with addition of init-sync weak block providers scores), so keeping those within status object will clutter interface even more. Do you think that I still should merge scorer and status together? One possible, and more clear imo, implementation can be to introduce Peers object, that will keep track of peers data (and pruning), and both Status and PeerScorer can rely on that. Thoughts?
>
>My point is that: bad responses functionality tracking shouldn't have been included into status object in the first place. That status object can, of course, use scorer or similar component to track and report performance, but its main responsibility seems to give peer connectivity status: connected, disconnected etc